lib/Dialect/Moore/Transforms/MergeAssignments.cpp

```diff
@@ -0,0 +1,79 @@
+//===- MergeAssignments.cpp - Merge declaration and assignment ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the MergeAssignments pass.
+// Find easy uses of declaration and assignments and merge them into
+// assigned_variable. Easy use represents the declaration doesn't be performed
+// bit slice. Like bit [127:0] b; assign b = 128'b0;
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+#include "mlir/IR/Dialect.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_MERGEASSIGNMENTS
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct MergeAssignmentsPass
+    : public circt::moore::impl::MergeAssignmentsBase<MergeAssignmentsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass> circt::moore::createMergeAssignmentsPass() {
+  return std::make_unique<MergeAssignmentsPass>();
+}
+
+// TODO: The net can be driven multiple times. However, the related rule is
+// complicated. So we will implement this in the future.
+
+// Only collect the easy declaration and its value at module level.
+static void collectAssignmets(SVModuleOp moduleOp,
+                              DenseMap<Value, Value> &assignments) {
+  moduleOp->walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op).Case<ContinuousAssignOp>([&](auto op) {
+      if (isa<VariableOp, NetOp>(op.getDst().getDefiningOp())) {
+        if (!assignments.lookup(op.getDst()))
+          assignments[op.getDst()] = op.getSrc();
+        else
+          mlir::emitError(op.getLoc())
+              << "Unsupported drive the same wire '"
+              << op.getDst().template getDefiningOp<NetOp>().getName()
+              << "' two times now";
+      }
+    });
+  });
+}
+
+void MergeAssignmentsPass::runOnOperation() {
+  OpBuilder builder(&getContext());
+
+  // Use to collect the easy declaration and its value.
+  DenseMap<Value, Value> assignments;
+
+  collectAssignmets(getOperation(), assignments);
+  for (auto assignment : assignments) {
+    auto varName =
+        assignment.first.getDefiningOp()->getAttrOfType<StringAttr>("name");
+
+    builder.setInsertionPointAfterValue(assignment.first);
+    builder.create<AssignedVarOp>(assignment.first.getLoc(),
+                                  assignment.first.getType(), varName,
+                                  assignment.second);
```

Comment:
I think this might turn a `NetOp` into an `AssignedVarOp` :thinking:, because you are collecting `NetOp` and `VariableOp`s into the `assignments` map.

Reply:
Yeah, `NetOp` also is merged into `AssignedVarOp`. What you meant was only merge `variable` and `assignment` into assigned_variable, and don't handle the `net` and its `assignment`, right?

Reply:
Yes exactly, we shouldn't change nets into variables or vice versa. I think it's still valuable to look at `NetOp`s and continuous assignments to them and if there is one single unique assignment to fold that into the `NetOp` itself, otherwise we'll just leave it as it is. That could be its own canonicalization though 😃.

Reply:
Let me attempt to add this canonicalization to it.